Create rule S6309: Functions returning Flow/Channel should not be suspending (#175)
SonarSource/rspec: rules/S6309/kotlin/rule.adoc

> In asynchronous programming in Kotlin with coroutines there are 2 ways of defining an async function. The first one is using modifier `suspend` in the function declaration and the second one is to create an extension function on `CoroutineScope`.

Reviewer: There are two ways to define asynchronous functions in Kotlin:
* using the modifier `suspend` in the function declaration
* creating an extension function on `CoroutineScope`

> `suspend` modifier is used for the functions that might take some time to compute the result. Which means that they can potentially suspend the caller coroutine.

Reviewer: The `suspend` modifier is generally used for functions that might take some time to complete. The caller coroutine might be potentially suspended.

Reviewer 2: +1 to Alex's suggestion, one super minor suggestion for the suggestion :). switch "be" & "potentially":
"...might potentially be suspended"

> For the functions that return result immediately but are starting the coroutine in the background it is suggested to write an extension function for `CoroutineScope`.

Reviewer: For the functions that return results immediately but starts the coroutine in the background, it is suggested to write an extension function on `CoroutineScope`. This means that extension functions on `CoroutineScope`s are not supposed to be `suspend`ed.

> Functions returning `Flow` or `Channel` are supposed to return the result immediately and might want to start a new coroutine in the background. So we should not mark them as `suspend` and if there is a need to launch some coroutine in the background, it's recommended to make them extension functions on `CoroutineScope`. 

Reviewer: Functions returning `Flow` or `Channel` are supposed to return the result immediately and might want to start a new coroutine in the background. As q consequence, such functions should not be declared as `suspend`. If there is a need to launch some coroutine in the background, it's recommended to declare them as extension functions on `CoroutineScope`.

== Noncompliant Code Example

----
suspend fun f(): Flow<Int> {
    var flowCollector: FlowCollector<Int>? = null
    val flow = flow<Int> {
        flowCollector = this
    }

>     flowCollector?.emit(1)

Reviewer 2: I found this flowCollector variable a bit confusing when reading the example. Is there a way to simplify this or replace some logic with a comment along the lines of `// Do something`? I'm not super familiar with flows at this point, so perhaps this is necessary, but it seems a bit confusing.

    return flow
}

suspend fun f(): Channel<Int> {
    val ch = Channel<Int>()
    ch.send(1)
    return ch

> }

Reviewer 2: I'd suggest putting the 2 code examples into 2 different code boxes.

----

== Compliant Solution

----
fun f(): Flow<Int> = flow {
    emit(1)
}

fun CoroutineScope.f(): Channel<Int> {
    val ch = Channel<Int>()
    launch {
        ch.send(1)
    }
    return ch

> }

Reviewer 2: I'd suggest putting the 2 code examples into 2 different code boxes.

There are two ways to define asynchronous functions in Kotlin:

> * using the modifier suspend in the function declaration

Reviewer 2: ```suggestion
* using the modifier `suspend` in the function declaration
```

> * creating an extension function on CoroutineScope

Reviewer 2: ```suggestion
* creating an extension function on `CoroutineScope`
```

The `suspend` modifier is generally used for functions that might take some time to complete. The caller coroutine might potentially be suspended.

> For the functions that return results immediately but starts the coroutine in the background, it is suggested to write an extension function on `CoroutineScope`. This means that extension functions on `CoroutineScopes` are not supposed to be suspending.
> 
> Functions returning `Flow` or `Channel` are supposed to return the result immediately and might want to start a new coroutine in the background. As a consequence, such functions should not be suspending. If there is a need to launch some coroutine in the background, it's recommended to declare them as extension functions on `CoroutineScope`.

Reviewer 2: ```suggestion
Functions that start a coroutine in the background and return before said coroutine has completed running should be extension functions on `CoroutineScope`. This helps to clarify the intention of such a function. Further, such such functions should not be suspending, as suspending functions should only return once all the work they are designed to perform is complete.

Functions returning `Flow` or `Channel` should return the result immediately and may start a new coroutine in the background. As a consequence, such functions should not be suspending and if they launch a coroutine in the background, they should be declared as extension functions on `CoroutineScope`.
```
